### PR TITLE
reimplement cityscapes

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR reimplemented Cityscapes dataset support.
1. Add script to easily convert original cityscapes dataset into coco json format. 
2. Evaluate results with Cityscapes official API.

### CODE REVIEW

**Comment by liyiwei979621500** [comment]
I don't think it a good idea to read Read original Cityscapes dataset annotation since it's very slow compared to using genertated coco format annotation(about 5s).
Fewer launch time is quiet important when implementing new model, during which developer frequently run trainning to see is all the tensor shapes all matched.

**Comment by xvjiarui** [comment]
> I don't think it a good idea to read Read original Cityscapes dataset annotation since it's very slow compared to using genertated coco format annotation(about 5s).
> Fewer launch time is quiet important when implementing new model, during which developers frequently run trainning to see if all the tensor shapes all matched.

Thanks for your valuable advice. 

Actually, I just updated the comment of this PR. A user-friendly convert script is provided to generate COCO style annotations. You could check it out [here](https://github.com/xvjiarui/mmdetection/blob/cityscapes/tools/convert_datasets/cityscapes.py). It would be merged into master recently. 

Btw, if there is any issue about this PR, you are very welcomed to point it out or make a PR in the future.

**Comment by liyiwei979621500** [comment]
Nice! Looking forward to the new implementation:)

**Comment by liyiwei979621500** [comment]
Wiil the cityscapes class order affect mAP?

**Comment by liyiwei979621500** [comment]
Is the mAP in new md file AP0.5 or AP0.5:0.95?

File: mmdet/datasets/cityscapes.py [hunk 1]
```diff
@@ -1,9 +1,277 @@
...
 from .registry import DATASETS
 
 
 @DATASETS.register_module
 class CityscapesDataset(CocoDataset):
-
+    """
+    The correct CLASSES order in official cityscapes should be
+    CLASSES = ('person', 'rider', 'car', 'truck', 'bus', 'train', 'motorcycle',
+                'bicycle')
```
Comment (hellock) [inline mmdet/datasets/cityscapes.py]: This comment can be removed.

File: tools/convert_datasets/cityscapes.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    args = parse_args()
+    cityscapes_path = args.cityscapes_path
+    out_dir = args.out_dir if args.out_dir else cityscapes_path
+    mmcv.mkdir_or_exist(out_dir)
+
+    img_dir = osp.join(cityscapes_path, args.img_dir)
+    gt_dir = osp.join(cityscapes_path, args.gt_dir)
+
+    set_name = dict(
+        train='instance_train_gtFine.json',
```
Comment (hellock) [inline tools/convert_datasets/cityscapes.py]: We may also use `instancesonly_filtered_gtFine_xxx.json`

File: mmdet/datasets/cityscapes.py [hunk 3]
```diff
@@ -1,9 +1,277 @@
...
+            tmp_dir = None
+        result_files = self.results2txt(results, txtfile_prefix)
+
+        return result_files, tmp_dir
+
+    def evaluate(self,
+                 results,
+                 metric='bbox',
+                 logger=None,
+                 jsonfile_prefix=None,
```
Comment (hellock) [inline mmdet/datasets/cityscapes.py]: Maybe renamed to `outfile_prefix` since the output file may be a json or txt.

File: mmdet/datasets/cityscapes.py [hunk 4]
```diff
@@ -1,9 +1,277 @@
...
+            iou_thrs (Sequence[float]): IoU threshold used for evaluating
+                recalls. If set to a list, the average recall of all IoUs will
+                also be computed. Default: 0.5.
+
+        Returns:
+            dict[str: float]
+        """
+        eval_results = dict()
+
+        metrics = metric if isinstance(metric, list) else [metric]
```
Comment (hellock) [inline mmdet/datasets/cityscapes.py]: Copy it if it is a list, because we remove `cityscapes` from the list.

File: configs/cityscapes/README.md [hunk 5]
```diff
@@ -1,28 +1,21 @@
...
 - 1x training schedule indicates 64 epochs which corresponds to slightly less than the 24k iterations reported in the original schedule from the [Mask R-CNN paper](https://arxiv.org/abs/1703.06870)
-- All pytorch-style pretrained backbones on ImageNet are from PyTorch model zoo.
-
-
-## Baselines
-
-Download links and more models with different backbones and training schemes will be added to the model zoo.
+- COCO pre-trained weights are used to initialize.
+- A conversion [script](../../tools/convert_datasets/cityscapes.py) is provided to convert Cityscapes into COCO format. Please refer to [INSTALL.md](../../docs/INSTALL.md#prepare-datasets) for details.
+- `CityscapesDataset` implemented three evaluation methods. `bbox` and `segm` are standard COCO bbox/mask AP. `cityscapes` is the cityscapes dataset official evaluation, which may be slightly higher than COCO.
```
Comment (hellock) [inline configs/cityscapes/README.md]: Add an example how to use the official cityscapes evaluation protocol and how to prepare results for submission to the evaluation server in `GETTING_STARTED.md`.

File: tools/convert_datasets/cityscapes.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            copyfile(img_file, osp.join(target_dir, osp.basename(img_file)))
+            copyfile(segm_file, osp.join(target_dir, osp.basename(segm_file)))
+        else:
+            move(img_file, osp.join(target_dir, osp.basename(img_file)))
+            move(segm_file, osp.join(target_dir, osp.basename(segm_file)))
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description='Convert Cityscapes annotations to mmdetection format')
```
Comment (hellock) [inline tools/convert_datasets/cityscapes.py]: to COCO format

File: configs/cityscapes/README.md [hunk 7]
```diff
@@ -1,28 +1,21 @@
...
 
 
 ### Faster R-CNN
 
 |    Backbone     |  Style  | Lr schd | Scale    | Mem (GB) | Train time (s/iter) | Inf time (fps) | box AP | Download |
 | :-------------: | :-----: | :-----: | :---:    | :------: | :-----------------: | :------------: | :----: | :------: |
-|    R-50-FPN     | pytorch |   1x    | 800-1024 | 4.9      | 0.345               | 8.8            | 36.0   | [model](https://open-mmlab.s3.ap-northeast-2.amazonaws.com/mmdetection/models/cityscapes/faster_rcnn_r50_fpn_1x_city_20190727-7b9c0534.pth) |
+|    R-50-FPN     | pytorch |   1x    | 800-1024 | 4.9      | 0.345               | 8.8            | 41.6   | [model](https://open-mmlab.s3.ap-northeast-2.amazonaws.com/mmdetection/models/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes_20200227-362cfbbf.pth) |
```
Comment (hellock) [inline configs/cityscapes/README.md]: Train time and Inf time are outdated.

File: configs/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # model settings
 model = dict(
     type='FasterRCNN',
-    pretrained='modelzoo://resnet50',
+    pretrained='torchvision://resnet50',
```
Comment (hellock) [inline configs/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes.py]: Since we have specified `load_from`, `pretrained` can be left None.

**Comment by liyiwei979621500** [comment]
The tools/test.py is running on the test set by default.
image

https://github.com/open-mmlab/mmdetection/blob/51df8a9b7ad5f25ebd75cf8e0969c3b728bde08d/configs/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes.py#L149-L154

Should it be changed to val?

**Comment by xvjiarui** [comment]
> The tools/test.py is running on the test set by default.
> image
> 
> https://github.com/open-mmlab/mmdetection/blob/51df8a9b7ad5f25ebd75cf8e0969c3b728bde08d/configs/cityscapes/faster_rcnn_r50_fpn_1x_cityscapes.py#L149-L154
> 
> Should it be changed to val?

We usually submit `test` set results of Cityscapes instance segmentation. You could change it to `val`  set in your program if needed.
